components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/CacheBackedActionMgtDAO.java

```diff
@@ -124,9 +127,51 @@ public Map<String, Integer> getActionsCountPerType(Integer tenantId) throws Acti
     }
 
     @Override
-    public Action getActionByActionId(String actionId, Integer tenantId) throws ActionMgtException {
+    public Action getActionByActionId(String actionType, String actionId, Integer tenantId) throws ActionMgtException {
 
-        return actionManagementDAO.getActionByActionId(actionId, tenantId);
+        ActionTypeCacheKey cacheKey = new ActionTypeCacheKey(actionType);
+        ActionCacheEntry entry = actionCacheByType.getValueFromCache(cacheKey, tenantId);
+        List<Action> actionsFromCache = new ArrayList<>();
+
+        /* If the entry for the given action type is not null, get the action list from cache and iterate to get the
+         action by matching action id. */
+        if (entry != null) {
+            actionsFromCache = entry.getActions();
+            for (Action action: actionsFromCache) {
+                if (StringUtils.equals(action.getId(), actionId)) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Action is found from the cache with action Id " + actionId);
+                    }
+                    return action;
+                }
+            }
+        }
```

Comment:
better extract to a different method.

Reply:
If we are extracting this to separate method, only the for loop can be extracted.